code/processes/segmentedtickerplant.q

```diff
@@ -0,0 +1,124 @@
+    .proc.loadf[schemafile:raze .proc.params[`schemafile]];
+    [.lg.e[`stp;"schema file required"];exit 1]
+   ]
+  ];
+
+// In none or tabular mode, intraday rolling not required
+if[.stplg.multilog in `none`tabular;.stplg.multilogperiod:1D];
+
+// In custom mode, load logging type for each table
+if[.stplg.multilog~`custom;
+  @[{.stplg.custommode:1_(!) . ("SS";",")0: x};.stplg.customcsv;
+    {.lg.e[`stp;"failed to load custom mode csv"]}]
+ ];
+
+// functions used by subscribers
+tablelist:{.stpps.t}
+// subscribers who want to replay need this info 
+subdetails:{[tabs;instruments]
+ `schemalist`logfilelist`rowcounts`date`logdir!(.u.sub\:[tabs;instruments];.stplg.replaylog[tabs];tabs#.stplg `rowcount;(.eodtime `d);`$getenv`KDBSTPLOG)}
+
+// Generate table and schema information and set up default table UPD functions
+generateschemas:{
+  .stpps.init[];
+  $[.sctp.chainedtp;
+    .stplg.updtab:(.stpps.t!(count .stpps.t)#{[x;y] x}),.stplg.updtab;
+    .stplg.updtab:(.stpps.t!(count .stpps.t)#{(enlist(count first x)#y),x}),.stplg.updtab
+    ]
+  }
+
+// Initialise UPD and ZTS behaviour based on batching mode (b)
+init:{[b]
+  if[not all b in/:(key .stplg.upd;key .stplg.zts);'"mode ",(string b)," must be defined in both .stplg.upd and .stplg.zts"];
+  .stplg.updmsg:.stplg.upd[b];
+  $[.sctp.chainedtp;
+    .u.upd:{[t;x]
+      now:.z.p;
+      // Type check allows update messages to contain multiple tables/data
+      $[0h<type t;
+        .stplg.updmsg'[t;x;now+.eodtime.dailyadj];
+        .stplg.updmsg[t;x;now+.eodtime.dailyadj]
+      ];
+      .stplg.seqnum+:1;
+      };
+    .u.upd:{[t;x]
+      // snap the current time and check for period end
+      if[.stplg.nextendUTC<now:.z.p;.stplg.checkends now];
+      // Type check allows update messages to contain multiple tables/data
+      $[0h<type t;
+        .stplg.updmsg'[t;x;now+.eodtime.dailyadj];
+        .stplg.updmsg[t;x;now+.eodtime.dailyadj]
+      ];
+      .stplg.seqnum+:1;
+      }
+    ];
+  // set .z.ts to execute the timer func and then check for end-of-period/end-of-day
+  .stplg.ts:.stplg.zts[b];
+  $[.sctp.chainedtp;
+    .z.ts:{
+      .stplg.ts now:.z.p;
+      };
+    .z.ts:{
+      .stplg.ts now:.z.p; 
+      .stplg.checkends now
+      }
+    ];
+  
+  // Error mode - write failed updates to separate TP log
+  if[.stplg.errmode;
+    //.stplg.openlogerr[.stplg.dldir]; - this is being done in .stplg.init now
+    .stp.upd:.u.upd;
+    .u.upd:{[t;x] .[.stp.upd;(t;x);{.stplg.badmsg[x;y;z]}[;t;x]]}
+  ];
+  // default the timer if not set
+  if[not system"t"; 
+   .lg.o[`timer;"defaulting timer to 1000ms"];
+    system"t 1000"];
+ };
+
+// Initialise process
+
+\d .
+
+// Set update and publish modes
+init[.stplg.batchmode]
+
+// subscribe to segmented tickerplant is mode is turned on
+if[.sctp.chainedtp;
+  endofperiod:{[x;y;z] .stplg.endofperiod[x;y;z]};
+  endofday:{[x;y] .stplg.endofday[x;y]};
+  .servers.startup[]; 
+  .sctp.subscribe[]
+  ]
+
```

Comment:
we probably need a .z.exit defined in here to close up the log files (write the meta data table correctly) on clean exit. It needs associated tests too

Reply:
This is now defined in stplog.q and there are tests in tests/stp/exit